components/lock/mongo/mongo_lock.go

```diff
@@ -0,0 +1,250 @@
+	UNLOCK_FAIL             = 6
+)
+
+// mongo lock store
+type MongoLock struct {
+	factory utils.MongoFactory
+
+	client     utils.MongoClient
+	session    utils.MongoSession
+	collection utils.MongoCollection
+	metadata   utils.MongoMetadata
+
+	features []lock.Feature
+	logger   log.ErrorLogger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewMongoLock returns a new mongo lock
+func NewMongoLock(logger log.ErrorLogger) *MongoLock {
+	s := &MongoLock{
+		features: make([]lock.Feature, 0),
+		logger:   logger,
+	}
+	return s
+}
+
+func (e *MongoLock) Init(metadata lock.Metadata) error {
+	var client utils.MongoClient
+	// 1.parse config
+	m, err := utils.ParseMongoMetadata(metadata.Properties)
+	if err != nil {
+		return err
+	}
+	e.metadata = m
+
+	e.factory = &utils.MongoFactoryImpl{}
+
+	// 2. construct client
+	if client, err = e.factory.NewMongoClient(m); err != nil {
+		return err
+	}
+
+	e.ctx, e.cancel = context.WithCancel(context.Background())
+
+	if err := client.Ping(e.ctx, nil); err != nil {
+		return err
+	}
+
+	wc, err := utils.GetWriteConcernObject(e.metadata.WriteConcern)
+	if err != nil {
+		return err
+	}
+
+	rc, err := utils.GetReadConcrenObject(e.metadata.ReadConcern)
+	if err != nil {
+		return err
+	}
+
+	// set mongo options of collection
+	opts := options.Collection().SetWriteConcern(wc).SetReadConcern(rc)
+
+	// create database
+	database := client.Database(e.metadata.DatabaseName)
+
+	// create collection
+	e.collection = e.factory.NewMongoCollection(database, e.metadata.CollectionName, opts)
+
+	// create exprie time index
+	indexModel := mongo.IndexModel{
+		Keys:    bsonx.Doc{{"Expire", bsonx.Int64(1)}},
+		Options: options.Index().SetExpireAfterSeconds(0),
+	}
+	e.collection.Indexes().CreateOne(e.ctx, indexModel)
+
+	e.client = client
+
+	return err
+}
+
+// Features is to get MongoLock's features
+func (e *MongoLock) Features() []lock.Feature {
+	return e.features
+}
+
+func (e *MongoLock) TryLock(req *lock.TryLockRequest) (*lock.TryLockResponse, error) {
+	var err error
+	// create mongo session
+	e.session, err = e.client.StartSession()
+	txnOpts := options.Transaction().SetReadConcern(readconcern.Snapshot()).
+		SetWriteConcern(writeconcern.New(writeconcern.WMajority()))
+
+	// close mongo session
+	defer e.session.EndSession(e.ctx)
+
+	// start transaction
+	status, err := e.session.WithTransaction(e.ctx, func(sessionContext mongo.SessionContext) (interface{}, error) {
+		var err error
+		var singleResult *mongo.SingleResult
+		var insertOneResult *mongo.InsertOneResult
+
+		// set exprie date
+		expireTime := time.Now().Add(time.Duration(req.Expire) * time.Second)
+
+		// find mongo lock
+		singleResult = e.collection.FindOne(e.ctx, bson.M{"_id": req.ResourceId})
+
+		// insert mongo lock
+		if singleResult.Err() == mongo.ErrNoDocuments {
+			insertOneResult, err = e.collection.InsertOne(e.ctx, bson.M{"_id": req.ResourceId, "LockOwner": req.LockOwner, "Expire": expireTime})
+		}
```

Comment:
Is it necessary to find first and then insert ？ only insert  and if  `_id` exsit ,  mongo will return error？

Reply:
yes, primary key unique

Reply:
so the transaction maybe not necessary ？